### DISCUSSION TITLE
importing .eaf's that deviate from ACLEW DAS

### DISCUSSION THREAD

**Discussion by alecristia** [LAAC-LSCP/ChildProject#204] [Q&A] [answered]
Most of our .eaf's will follow the ACLEW DAS: One tier, one speaker.

But the ones I'm looking at have only one tier (see png2019/annotations/mc/raw): the speaker ID is the content of each segment, and there is a dependent tier with the transcription of what has been said.

I'd like to know whether the best strategy is to first convert this unusual format to a more usual format, then import; or whether I should be handling this with some command within the annotation manager; or perhaps there is a third option that I'm not thinking about.

Thank you in advance!

**Comment by lucasgautheron** [comment, marked as answer]
The best strategy in my opinion would be to use the python API of ChildProject (`ChildProject.annotations.AnnotationManager.import_annotations`) with a custom converter.

A converter is any function that takes the path to an annotation as an input and returns a dataframe of all segments, according to the format defined here: https://childproject.readthedocs.io/en/latest/format.html#annotations-format

This is the converter that is used for ACLEW eaf: https://github.com/LAAC-LSCP/ChildProject/blob/master/ChildProject/annotations.py#L252

Using a custom converter would look like this:

```python
def converter(filename):
    df = pd.DataFrame()
    eaf = pympi.Elan.Eaf(filename)
    # ...
    return df

project = ChildProject('.')
am = AnnotationManager(project)
am.import_annotations(input = eaf_to_import, import_function = converter)
```

where `eaf_to_import` is a dataframe of annotations to import as defined in https://childproject.readthedocs.io/en/latest/format.html#annotation-importation-input-format.

So that requires a bit of work with the pympi package, but I think this is the cleanest way possible.

**Reply by lucasgautheron** [reply]
@sarpu @alecristia 
This might be useful: https://childproject.readthedocs.io/en/latest/api-annotations.html

**Comment by alecristia** [comment]
Thank you, that helps! 

I've a couple more questions:

1. While trying to create the right importer, I ran it a few times, and I think it added lines to annotations.csv because:

`$ datalad status`
> modified: metadata/annotations.csv (file)
 modified: scripts/import_eaf.py (file)

Even after I clean up the wrong ones with:

`$ child-project remove-annotations . --set mc`

(no errors or warnings ensue)

I still get:

`$ datalad status`
> modified: metadata/annotations.csv (file)
 modified: scripts/import_eaf.py (file)

2. The second question is one of sheer ignorance -- there is something I'm not quite getting as to how things work. I think my script for eaf import is by and large getting all the info I need -- there is just one bug to correct (the fact that transcriptions == "CRYING" need to result in vcm == "C", whereas now VCM is always NA). However, it creates empty csvs. I'm not attaching a minimally reproducible example, as the paths will not be respected given that I'm taking the files out of the ChildProject hierarchy & I'm concerned about sharing files with transcriptions in them. However, here is what my script looks like:

```
#!/usr/bin/env python3
from ChildProject.projects import ChildProject
from ChildProject.annotations import AnnotationManager
import os
import pandas as pd
import glob
import pympi

project = ChildProject('.')
am = AnnotationManager(project)

files = pd.DataFrame([
    {'raw_filename': os.path.basename(f)}
    for f in glob.glob(os.path.join('annotations/mc/raw/*.eaf'))
])



extract = files['raw_filename'].str.extract(r"(c[0-9]+)_([0-9]{8})_([0-9]{4})")
files['child_id'] = extract[0]
files = files.merge(project.recordings, how = 'inner', left_on = 'child_id', right_on = 'child_id')


children = pd.read_csv('extra/humanval/seek_times.csv', dtype = {'child_id': str})
files = files.merge(children, how = 'inner', left_on = 'child_id', right_on = 'child_id')



files['time_seek'] = (files['on'])*1000
files['set'] = "mc"

input = files[['recording_filename_x', 'raw_filename', 'time_seek', 'set']]
input.rename(columns = {
    'recording_filename_x': 'recording_filename'
}, inplace = True)

#only one minute, from 120 secs to 180 secs, has been annotated
input['format'] = 'eaf'
input['range_onset'] = "120"
input['range_offset'] = "180"
#input.to_csv('test.csv')

SPEAKER_ID_TO_TYPE = {
        'OC': 'OCH',
        'TC': 'CHI',
        'FA': 'FEM',
        'MA': 'MAL'
    }

TRANSCRIPTION_TO_VCM = {
        'CRYING.': 'C',
        'CRYING': 'C',
        'SCREAM.': 'C',
        'SCREAM': 'C'
    }

def converter(filename):
    df = pd.DataFrame()
    eaf = pympi.Elan.Eaf(filename)
    segments = {}
#ignore all tiers except for the one called "recoding"
    annotations = eaf.tiers["recoding"][0]
# go through each of the annotations and log everything
    for aid in annotations:
        (start_ts, end_ts, value, svg_ref) = annotations[aid]
# skip the "do not annotate" segments
        if value != "DO NOT ANNOTATE":
           (start_t, end_t) = (eaf.timeslots[start_ts], eaf.timeslots[end_ts])
# note that we leave everything else that has been annotated in dependent tiers with NA -- it'll be filled in below
           segment = {
                    'segment_onset': int(round(start_t)),
                    'segment_offset': int(round(end_t)),
                    'speaker_id': value,
                    'speaker_type': SPEAKER_ID_TO_TYPE[value] if value in SPEAKER_ID_TO_TYPE else 'NA',
                    'vcm_type': 'NA',
                    'transcription': 'NA'
           }
           segments[aid] = segment
#next  we do the transcription tier
    reference_annotations = eaf.tiers["transcription"][1]

    for aid in reference_annotations:
        (ann, value, prev, svg) = reference_annotations[aid]

        ann = aid
        parentTier = eaf.tiers[eaf.annotations[ann]]
        while 'PARENT_REF' in parentTier[2] and parentTier[2]['PARENT_REF'] and len(parentTier[2]) > 0:
           ann = parentTier[1][ann][0]
           parentTier = eaf.tiers[eaf.annotations[ann]]

           if ann not in segments:
             print("warning: annotation '{}' not found in segments for '{}'".format(ann, filename))
             continue
                
           segment = segments[ann]

           segment['vcm_type'] = TRANSCRIPTION_TO_VCM[value] if value in TRANSCRIPTION_TO_VCM else 'NA'
           segment['transcription'] = value
    print(segments)
    return pd.DataFrame(segments.values())

project = ChildProject('.')
am = AnnotationManager(project)
am.import_annotations(input = input, import_function = converter)

#am.import_annotations(input)

```

And sample output produced for that file, which suggests to me that it IS getting the right info out of the eaf:

> 'a58': {'segment_onset': 135863, 'segment_offset': 136411, 'speaker_id': '2T', 'speaker_type': 'NA', 'vcm_type': 'NA', 'transcription': ''}, 'a59': {'segment_onset': 136418, 'segment_offset': 137091, 'speaker_id': 'OC', 'speaker_type': 'OCH', 'vcm_type': 'NA', 'transcription': 'SINGING.'},

**Reply by lucasgautheron** [reply]
Regarding your first question, you can still fall back on your feet with `git reset --hard` and `git restore metadata/annotations.csv` (I suggest you do that)

Regarding your second question, I think the problem lies here:

```python
input['range_onset'] = "120"
input['range_offset'] = "180"
```

It should be:

```python
input['range_onset'] = 120*1000
input['range_offset'] = 180*1000
```

(Times should be expressed in milliseconds). The reason is the importer[ will remove segments that don't lie in this range](https://github.com/LAAC-LSCP/ChildProject/blob/master/ChildProject/annotations.py#L581). So if the range is set to [120 ms , 180 ms] then all the segments will be dropped and the CSV will be empty.

**Reply by alecristia** [reply]
thank you, both things helped!

I probably have another bug, because I get these warnings; notice that the segment numbers are repeated (this is a partial list, the segments a22 a 42 etc get repeated for each eaf):

> warning: annotation 'a22' not found in segments for './annotations/mc/raw/c23_20190822_1255.eaf'
warning: annotation 'a42' not found in segments for './annotations/mc/raw/c23_20190822_1255.eaf'
warning: annotation 'a48' not found in segments for './annotations/mc/raw/c23_20190822_1900.eaf'
warning: annotation 'a93' not found in segments for './annotations/mc/raw/c23_20190822_1900.eaf'
warning: annotation 'a46' not found in segments for './annotations/mc/raw/c23_20190822_2015.eaf'
warning: annotation 'a90' not found in segments for './annotations/mc/raw/c23_20190822_2015.eaf'
warning: annotation 'a59' not found in segments for './annotations/mc/raw/c23_20190822_1340.eaf'
warning: annotation 'a116' not found in segments for './annotations/mc/raw/c23_20190822_1340.eaf'
warning: annotation 'a59' not found in segments for './annotations/mc/raw/c23_20190822_1200.eaf'
warning: annotation 'a116' not found in segments for './annotations/mc/raw/c23_20190822_1200.eaf'
warning: annotation 'a3' not found in segments for './annotations/mc/raw/c23_20190822_1305.eaf'
warning: annotation 'a4' not found in segments for './annotations/mc/raw/c23_20190822_1305.eaf'
warning: annotation 'a22' not found in segments for './annotations/mc/raw/c23_20190822_1255.eaf'
warning: annotation 'a42' not found in segments for './annotations/mc/raw/c23_20190822_1255.eaf'
warning: annotation 'a46' not found in segments for './annotations/mc/raw/c23_20190822_2015.eaf'
warning: annotation 'a48' not found in segments for './annotations/mc/raw/c23_20190822_1900.eaf'
warning: annotation 'a90' not found in segments for './annotations/mc/raw/c23_20190822_2015.eaf'
warning: annotation 'a93' not found in segments for './annotations/mc/raw/c23_20190822_1900.eaf'
warning: annotation 'a59' not found in segments for './annotations/mc/raw/c23_20190822_1340.eaf'
warning: annotation 'a116' not found in segments for './annotations/mc/raw/c23_20190822_1340.eaf'
warning: annotation 'a59' not found in segments for './annotations/mc/raw/c23_20190822_1200.eaf'
warning: annotation 'a116' not found in segments for './annotations/mc/raw/c23_20190822_1200.eaf'

**Reply by alecristia** [reply]
hmm I'm afraid I have more bugs. You can let me know if this is beyond the scope of the help one should get in this discussion!

- my code seems to be carrying over lines across different files that belong to the same child, but not all of them:

       11 annotations/mc/converted/e20190821_134438_013100_2130000.0_120000.csv
      73 annotations/mc/converted/e20190821_134438_013100_22230000.0_120000.csv
      73 annotations/mc/converted/e20190821_134438_013100_24930000.0_120000.csv
      73 annotations/mc/converted/e20190821_134438_013100_32430000.0_120000.csv
      73 annotations/mc/converted/e20190821_134438_013100_34530000.0_120000.csv
      73 annotations/mc/converted/e20190821_134438_013100_5130000.0_120000.csv
      77 annotations/mc/converted/e20190822_153540_013103_1248000.0_120000.csv
      77 annotations/mc/converted/e20190822_153540_013103_24648000.0_120000.csv
      77 annotations/mc/converted/e20190822_153540_013103_48000.0_120000.csv
      77 annotations/mc/converted/e20190822_153540_013103_6348000.0_120000.csv
      77 annotations/mc/converted/e20190822_153540_013103_8448000.0_120000.csv
      77 annotations/mc/converted/e20190822_153540_013103_9348000.0_120000.csv
      49 annotations/mc/converted/e20190823_142220_009458_3640000.0_120000.csv
      49 annotations/mc/converted/e20190823_142220_009458_38140000.0_120000.csv
      49 annotations/mc/converted/e20190823_142220_009458_39340000.0_120000.csv
      49 annotations/mc/converted/e20190823_142220_009458_40840000.0_120000.csv
      49 annotations/mc/converted/e20190823_142220_009458_41740000.0_120000.csv
      70 annotations/mc/converted/e20190823_142220_009458_42040000.0_120000.csv
      57 annotations/mc/converted/e20190823_150227_009456_10709670.0_120000.csv
      57 annotations/mc/converted/e20190823_150227_009456_29909670.0_120000.csv
      57 annotations/mc/converted/e20190823_150227_009456_34409670.0_120000.csv
      44 annotations/mc/converted/e20190823_150227_009456_4709670.0_120000.csv
      57 annotations/mc/converted/e20190823_150227_009456_8009670.0_120000.csv

When I open `annotations/mc/converted/e20190821_134438_013100_24930000.0_120000.csv`
and `annotations/mc/converted/e20190821_134438_013100_22230000.0_120000.csv` the contents are identical

- I still haven't figured out why this line doesn't work: vcm_type is always NA, even though some lines do consist of "CRYING". I'm also not sure why those lines don't even appear in the transcription. (that is, lines where the transcription is CRYING don't get a non-NA vcm_type AND are removed from the log)
`segment['vcm_type'] = TRANSCRIPTION_TO_VCM[value] if value in TRANSCRIPTION_TO_VCM else 'NA'`

**Reply by lucasgautheron** [reply]
I'll try to reproduce
Feel free to commit and push your work on a separate branch in the meantime:

```bash
git checkout -b eaf
datalad save . -m "eaf branch"
datalad run-procedure setup --confidential --private
datalad push --to private
datalad push --to confidential
datalad push --to el1000
```

**Reply by alecristia** [reply]
I got:
`$ datalad push --to el1000                                            ` 
> [ERROR  ] Unknown push target 'el1000'. Known targets: 'private', 'origin', 'confidential'. [publish(/Users/acristia/Documents/git-data/png2019)] 
publish(error): . [Unknown push target 'el1000'. Known targets: 'private', 'origin', 'confidential'.]

what did I do wrong?

**Comment by sarpu** [comment]
Just adding a comment here to have a record that we should include EE# (EE1, EE2, etc. ) in the output. Currently, annotations in those tiers don't seem to be imported.

**Reply by lucasgautheron** [reply]
> Just adding a comment here to have a record that we should include EE# (EE1, EE2, etc. ) in the output. Currently, annotations in those tiers don't seem to be imported.

Done! https://github.com/LAAC-LSCP/ChildProject/commit/fc20b275df971f05f0420f0a7ece931852767880
